HTML API: CSS matching

sirreal opened the pull request:
try with https://github.com/sirreal/html-api-debugger/pull/5

Trac ticket: <!-- insert a link to the WordPress Trac ticket here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
WIP class skeleton

sirreal committed:
Document class

sirreal committed:
Do not support namespaced selectors

sirreal committed:
Flesh out stuff

sirreal committed:
Starting to actually parse

sirreal committed:
Add ident tests

sirreal committed:
Fix ident non-ascii bug

sirreal committed:
Use class after defined

sirreal committed:
Fix some char stuff

sirreal committed:
Improve tests

sirreal committed:
Housekeeping

sirreal committed:
Require new file in WP

sirreal committed:
Fix offset type

sirreal committed:
Add more tests and invalid tests

sirreal committed:
Fix wrong offset var usage

sirreal committed:
comment tweak

sirreal committed:
Implement codepoint escape with strspn

sirreal committed:
Test with UPPER HEX

sirreal committed:
Add ID tests

sirreal committed:
Improve tests

sirreal committed:
Add class selector tests

sirreal committed:
Add class selector

sirreal committed:
Simplify id selector parse

sirreal committed:
Improve ident tests

sirreal committed:
Add type selector tests

sirreal committed:
Add docs and remove unreachable line

sirreal committed:
Add type selector class

sirreal committed:
Add attribute selector tests

sirreal committed:
improve attr tests

sirreal committed:
Fix expectation argument order

sirreal committed:
Add test and fix is_ident

sirreal committed:
Add parse_string stub

sirreal committed:
Add attribute selector parsing

sirreal committed:
Fix test expectations

sirreal committed:
More and improved attribute tests

sirreal committed:
Implement parse_string

sirreal committed:
Add string parse tests

sirreal committed:
Remove covers annotations

sirreal committed:
Remove unused line

sirreal committed:
Improve tests for 100% coverage on parse methods

sirreal committed:
Improve documentation

sirreal committed:
Fix parse return type and return annotations

sirreal committed:
Update documentation links and grammar

sirreal committed:
Update documentation and class name

sirreal committed:
Add selector class

sirreal committed:
Implement complex selector

sirreal committed:
Working and tested

sirreal committed:
Add select method

sirreal committed:
Merge branch 'html-api/add-css-selector-parser' into html-api/work-on-select-api

sirreal committed:
Selector parsing should allow cap I,S modifier

sirreal committed:
CSS Add matches to selector classes

sirreal committed:
Match is successful on _any_ match in selector list

sirreal committed:
ID matches depend on quirks mode

sirreal committed:
PICKME: Add is_quirks_mode method to processor

sirreal committed:
has_class may return null, coerce to bool

sirreal committed:
Update docs to only allow subclass selectors in final complex selector position

sirreal committed:
Restrict complex selectors to only allow subclass selectors in final position

sirreal committed:
Work on complex selector handling

sirreal committed:
Implement descendent selector matching

SantosGuillamot commented on src/wp-includes/html-api/class-wp-css-selectors.php:
```diff
@@ -0,0 +1,1169 @@
+				return false;
+			}
+		}
+		foreach ( $this->subclass_selectors as $subclass_selector ) {
+			if ( ! $subclass_selector->matches( $processor ) ) {
+				return false;
+			}
+		}
```
`$this->subclass_selectors` can be null, so it needs a safety check:

```php
if ( null !== $this->subclass_selectors ) {
	foreach ( $this->subclass_selectors as $subclass_selector ) {
		if ( ! $subclass_selector->matches( $processor ) ) {
			return false;
		}
	}
}
```

sirreal committed:
Add null check for subclass selectors

sirreal committed:
CSS selector reformat ternaries

sirreal committed:
Implement ~= attribute matching

sirreal committed:
DROPME: phpstan setup

sirreal committed:
CSS fix return type

sirreal committed:
Fix static analysis problems

sirreal committed:
Fix and annotate things (static analysis)

sirreal committed:
update tests

sirreal committed:
Id attribute must be a string to match id selector

sirreal committed:
Coerce boolean attributes to ""

sirreal committed:
Fix a few more static analysis things

sirreal committed:
Max phpstan level

swissspidy committed:
I18N: Load admin translations for auto update emails.

As a follow-up to [59460], make sure that admin strings are loaded when switching locales for auto update notification emails, as those strings are in a separate translation file.

Props benniledl, swissspidy.
Fixes #62496.

git-svn-id: https://develop.svn.wordpress.org/trunk@59478 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy committed:
Plugins: Make more plugin-related functions available early on.

This is a follow-up to [59461], which moved `get_plugin_data()` from `wp-admin/includes/plugin.php` to `wp-includes/functions.php` so it's available during the plugin loading process.

Related functions like `is_plugin_active()` are often used together and should therefore be moved as well, to improve backward compatibility for plugins which load `wp-admin/includes/plugin.php` only conditionally.

Props johnbillion, dd32, swissspidy.
See #62244.

git-svn-id: https://develop.svn.wordpress.org/trunk@59479 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Customize: Begin HTML markup before Customizer script hooks.

This prevents printing styles and scripts before the `<!DOCTYPE>`.

The `_wp_admin_html_begin()` function should precede Customizer script hooks, in case a plugin prints markup inside a hook such as `admin_enqueue_scripts`.

Follow-up to [19995], [27907].

Props sabernhardt.
Fixes #62629.

git-svn-id: https://develop.svn.wordpress.org/trunk@59480 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
External Libraries: Upgrade PHPMailer to version 6.9.3.

This is a maintenance release, adding support for the release version of PHP 8.4, and experimental support for PHP 8.5.

References:
* [https://github.com/PHPMailer/PHPMailer/releases/tag/v6.9.3 PHPMailer 6.9.3 release notes]
* [https://github.com/PHPMailer/PHPMailer/compare/v6.9.2...v6.9.3 Full list of changes in PHPMailer 6.9.3]

Follow-up to [50628], [50799], [51169], [51634], [51635], [52252], [52749], [52811], [53500], [53535], [53917], [54427], [54937], [55557], [56484], [57137], [59246].

Props desrosj, yogeshbhutkar, ayeshrajans.
Fixes #62632.

git-svn-id: https://develop.svn.wordpress.org/trunk@59481 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Unify parsing under single class

sirreal committed:
Rename files to align with class name

sirreal committed:
Add html processor select test suite

sirreal committed:
Fix select types

ockham committed:
Block Hooks: Fix context in `update_ignored_hooked_blocks_postmeta`.

Ensure that the `$context` arg passed from `update_ignored_hooked_blocks_postmeta` to `apply_block_hooks_to_content` (and from there, to filters such as `hooked_block_types` and `hooked_block`) has the correct type (`WP_Post`).

Filters hooked to `hooked_block_types` etc can typically include checks that conditionally insert a hooked block depending on `$context`. Prior to this changeset, a check like `if ( $context instanceof WP_Post )` would incorrectly fail, as `$context` would be a `stdClass` instance rather than a `WP_Post`. As a consequence, a hooked block inside of a Navigation post object that was modified by the user would not be marked as ignored by `update_ignored_hooked_blocks_postmeta`, and thus be erroneosly re-inserted by the Block Hooks algorithm.

Props bernhard-reiter.
Fixes #62639.

git-svn-id: https://develop.svn.wordpress.org/trunk@59482 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Update class doc

desrosj committed:
Build/Test Tools: Add `repository` input to support JSON reading workflow.

`actions/checkout` will always checkout the current repository unless the `repository` input is specified. This updates the `reusable-support-json-reader-v1.yml` workflow to always default to reading the JSON files from `wordpress-develop`.

A `repository` has also been added to the workflow to allow a different set of JSON files to be read if desired.

See #62221.

git-svn-id: https://develop.svn.wordpress.org/trunk@59483 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Improve select_ method arguments, docs, implementation

desrosj committed:
Build/Test Tools: Support older MariaDB versions in local Docker environment.

Older versions of MariaDB did not contain the `mariadb-admin` command. This command is configured as the `healthcheck` used by the local Docker environment to confirm that the database container has successfully started and is reporting as “healthy”. The current result is a failure when starting the environment while using one of the affected older versions.

For MariaDB versions 10.3 and earlier, the `mysqladmin` command was used instead. Since WordPress still technically supports back to MariaDB 5.5, the local environment should support running these versions. This updates the environment configuration to take this into account when performing a `healthcheck` test.

The README file is also updated to reflect that the same workaround added in [57568] for MySQL <= 5.7 is required when using MariaDB 5.5 on an Apple silicon machine.

Props johnbillion.
See #62221.

git-svn-id: https://develop.svn.wordpress.org/trunk@59484 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Run install tests when JSON reading workflow is changed.

Because the installation testing workflow relies on the reusable workflow that reads the JSON support files, it should be run when that file is changed to confirm there are no issues.

This is currently only configured for `pull_request` events, but should also be true for `push`.

See #62221.

git-svn-id: https://develop.svn.wordpress.org/trunk@59485 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Split classes into their own files

Satisfy the 1-class-per-file requirement

sirreal committed:
Remove redundant see phpdoc annotations

sirreal committed:
Fix docs and return type on select_all

sirreal committed:
Improve html select test docs

sirreal committed:
Add select support to tag processor

Split up main CSS selector class and support more restricted selectors in the tag processor.

sirreal committed:
DROPME: ignore code coverage HTML API

sirreal committed:
Simplify whitspace splitting function

sirreal committed:
Remove unreachable code

sirreal committed:
Add a lot of selector integration tests

sirreal committed:
Extract normalize input method

sirreal committed:
tests

sirreal committed:
Add nonfinal subclass selector test

SergeyBiryukov committed:
Docs: Remove blank line at the end of `wp_prepare_attachment_for_js()` DocBlock.

Follow-up to [21680], [49281].

Props nareshbheda.
Fixes #62642.

git-svn-id: https://develop.svn.wordpress.org/trunk@59486 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Fix logic bug in child selector exploration

sirreal committed:
Improve selector integration tests

swissspidy committed:
Plugins: Load `wp-admin/includes/plugin.php` earlier.

Partially reverts [59479] and [59461], which previously tried to move some functions from `wp-admin/includes/plugin.php` to `wp-includes/functions.php` so they are available early, so that `get_plugin_data()` can be used.

However, other functions from that file are often used by plugins without necessarily checking whether they are available, easily causing fatal errors. Requiring this file directly is a safer approach to avoid such errors.

Props peterwilsoncc, dd32, swissspidy, johnbillion.
Fixes #62244.

git-svn-id: https://develop.svn.wordpress.org/trunk@59488 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Try abstract class instead of interface

sirreal committed:
Revert "Try abstract class instead of interface"

This reverts commit 74881651faf991eabceb090707ce8b43c2a25316.

sirreal committed:
Clean up and document attribute selector

desrosj committed:
Build/Test Tools: Properly escape `$` characters in Docker compose file.

This fixes an invalid interpolation format error that can be encountered in the `mysql` container’s healthcheck test command.

Follow up to [59484].

Props afercia.
See #62221.

git-svn-id: https://develop.svn.wordpress.org/trunk@59489 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Use newer versions for `include` jobs.

The `include` part of the strategy for the PHPUnit testing workflow defines a few testing configurations outside of the matrix. The versions of PHP and MySQL used in these have not been updated for some time. This was mostly due to various incompatibilities that have since been resolved.

Props peterwilsoncc, johnbillion.
See #62221.

git-svn-id: https://develop.svn.wordpress.org/trunk@59490 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Support `trunk` as a version.

`trunk` is used interchangeably with `nightly`, so should be an accepted value when determining which version of WordPress is being tested.

Follow up to [59452], [59483].

Props johnbillion.
See #62221.

git-svn-id: https://develop.svn.wordpress.org/trunk@59491 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Introduce workflow for testing the local Docker environment.

While the PHPUnit workflow currently relies on the local Docker environment and provides some safety checks that the environment works as expected, this may not always be true and does not test all of the available commands related to the environment.

This introduces a basic workflow for testing the related scripts for the various supported combinations of PHP and database software with the environment to confirm everything is working as expected.

Ideally this would also be run on Windows and MacOS to catch platform specific bugs. Unfortunately, Docker is not supported within the GitHub Action runner images, so not all bugs will be caught by this workflow.

Props johnbillion, Clorith.
See #62221.

git-svn-id: https://develop.svn.wordpress.org/trunk@59492 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Update ticket number in tests

sirreal committed:
Improve some types

sirreal committed:
Fix and improve string token parsing

sirreal committed:
Update attribute selector tests

swissspidy committed:
Embeds: ensure correct thumbnail height.

Use height 0 instead of 9999 to avoid unnecessarily using the full size version.

Props colinleroy, swissspidy.
Fixes #62094.

git-svn-id: https://develop.svn.wordpress.org/trunk@59493 602fd350-edb4-49c9-b593-d223f7449a82

swissspidy committed:
I18N: Add new `WP_Locale::get_month_genitive()` method.

Complements existing helper methods such as `WP_Locale::get_month_abbrev()`.

Props ankitkumarshah, Tkama, SergeyBiryukov.
Fixes #58658.

git-svn-id: https://develop.svn.wordpress.org/trunk@59494 602fd350-edb4-49c9-b593-d223f7449a82

sirreal committed:
Revert "Update attribute selector tests"

This reverts commit 7df9ed91a1360d80c1dcb87980af941010b926ba.

sirreal committed:
Improve some complex selector match tests

sirreal commented:
Superseded by https://github.com/WordPress/wordpress-develop/pull/7857
